### ISSUE TITLE
Ability to manage manually registered client entries using Dynamic Client Registration Management protocol

### ISSUE THREAD

**Issue by aliaksander-samuseu** [GluuFederation/oxTrust#1908]
It seems like mostly web UI feature, but it also touches certain OIDC implementation nuances, so ties to oxAuth as well, thus may require attention from several developers.

As was requested by customer in ticket 7890, it would be useful to make manually registered OIDC clients optionally manageable via [Dynamic Registration Management protocol](https://tools.ietf.org/html/rfc7592#section-2.2). If this feature is enabled for a client, access token for its registration entry will be created and displayed on its page in web UI, for user to carry it over to the client out of band.

There is a possibility for this feature to synergy with oxAuth's "dynamicRegistrationEnabled" property quite nicely. This feature could be used as an per-client override when the former is set to "false", blocking dynamic registration, but still allowing them to remotely manage clients registered manually, what can be seen by many as a more secure approach (as proven by previous requests to protect the registration endpoint somehow).

FYI, while experimenting with it, I was able to turn on ability to view and update a manually registered client by adding next attributes to it:

 - `oxAuthRegistrationAccessToken` - some alpha-numeric string not longer than 35 characters; used to authorize your requests
 - `del` - set to "true"
 - `oxAuthClientIdIssuedAt` - set to date in recent past using next format `20200128231100.449Z` (seems like mandatory, oxAuth throws NPE otherwise)

I'll mention a few other thoughts and suggestions the customer made in the ticket:

1. They would like to have a way to change client's secret later via Dynamic Management protocol - AFAICR, it would violate the spec, according to [this RFC](https://tools.ietf.org/html/rfc7592#section-2.2), "The client MUST NOT be allowed to overwrite its existing client secret with its own chosen value"; @yuriyz  or @qbert2k should be able to comment on this

2. Proper logging should be implemented for this feature, showing clientID/name/etc auditing details whenever that token is used at INFO level (regardless of the correctness of the token value)

3. The password update functionality (in reality both for this and for all of gluu), should apply the rules of no t allowing dollar $ signs and any other characters that gluu does not handle well.

**Comment by mtgluu** [comment]
Thanks for the write up aliaksander. Yes it would be useful to be able to provide a registration_access_token for trusted/confidential clients created manually. I'm not sure if it should be tied exclusively to the "dynamicRegistrationEnabled" property though. It would seem better tied to a (new?) client attribute indicating whether the client was created manually or not. 

I went back a re-read the dynamic registration spec and there is some ambiguity in the way client_secrets are managed, even for dynamically registered clients. For example, you are correct in your response on 7890 that the spec states that the client must not be allowed to set their own secret value. However then later when discussing updates the spec states that:

> In section 2.2
   Upon successful update, the authorization server responds with an
   HTTP 200 OK message with content type "application/json" and a
   payload as described in Section 3.  Some values in the response,
   including the "client_secret" and "registration_access_token", MAY be
   different from those in the initial registration response.  If the
   authorization server includes a new client secret and/or registration
   access token in its response, the client MUST immediately discard its
   previous client secret and/or registration access token.  The value
   of the "client_id" MUST NOT change from the initial registration
   response.

Which seems to indicate then that the implementing OP may choose how/when to update the client secret whenever a client makes a request to the reg endpoint to make other updates. The expected functionality isn't really well documented in gluu though from the last time I looked. My thoughts are that whether it's a trusted/manually created client, or, a dynamically registered client, the client should be able to indicate whether or not gluu should update the client_secret value... even if it *is* generated by gluu instead of them (in order to be in compliance with the statement about the client NOT being able to set their own value)
